Updated `VolumeLevel` to use Decibel and Amplitude Ratio Units

bushrat011899 opened the pull request:
# Objective

- Fixes #9507

## Solution

- Breaking changes in both the `AudioSinkPlayback` `trait` and `VolumeLevel` type.
- Replaced `VolumeLevel` with an `enum` representing either the amplitude ratio or decibel level. Names and types were chosen to be consistent with 3rd party crates like `Kira`, and the internally used types in the rest of `bevy_audio`

---

## Changelog

- Moved `VolumeLevel` into its own file `volume_level.rs` to reduce clutter in `audio.rs`.
- Implemented total equality and ordering for `VolumeLevel`, as both variants `Amplitude` and `Decibels` are isomorphic and internally possess both properties.
- Implemented `amplitude()` and `decibels()` methods to get the desired unit type from `VolumeLevel` directly without a `match` statement.
- Added `debug_assert!` statements to ensure calculated values are within their appropriate bounds.
- Added a unit test for conversion between `Amplitude` and `Decibels`, with a truth table obtained from [Wikipedia](https://web.archive.org/web/20230810185300/https://en.wikipedia.org/wiki/Decibel)
- Updated internal functions to either accept the `VolumeLevel` type, or explicitly request an amplitude which is then wrapped in a `VolumeLevel::Amplitude` where appropriate.
- Removed `VolumeLevel::get` and `VolumeLevel::new` due to ambiguity in the unit to return.

## Migration Guide

- `VolumeLevel::new(volume)` -> `VolumeLevel::Amplitude(volume)`
- `VolumeLevel::get()` -> `VolumeLevel::amplitude()`
- `AudioSinkPlayback::volume()` -> `AudioSinkPlayback::volume().amplitude()`
- `AudioSinkPlayback::set_volume(volume)` -> `AudioSinkPlayback::set_volume(VolumeLevel::Amplitude(volume))`

killercup commented on crates/bevy_audio/src/volume_level.rs:
```diff
@@ -0,0 +1,154 @@
+pub enum VolumeLevel {
+    /// Volume Amplitude Ratio (unitless)
+    Amplitude(f32),
+    /// Decibels of Amplitude Ratio (dB)
+    Decibels(f32),
```
I can construct this with any valid float value, incl. NaN and Infinity. Maybe

- add the expected range to the doc comment
- or not make the variants public and add constructors
- or this normalized this on use and then warrants another doc comment :)

killercup replied:
Makes sense! From other Rust projects I'm used to quite high type safety so I was surprised to see this be unconstrained. Maybe an interesting idea to go thru the codebase and see where we can introduce some cheap constructors that make misuse less likely. I saw you already added a debug assert on output as well :)

rparrett commented on crates/bevy_audio/src/volume_level.rs:
```diff
@@ -0,0 +1,165 @@
+        };
+
+        debug_assert!(
+            (f32::MIN_POSITIVE..=f32::MAX).contains(&amplitude),
```
`0.0` seems like a perfectly valid amplitude to me.

bushrat011899 replied:
You've given me an idea, also resolved!

bushrat011899 commented:
[This comment](https://github.com/bevyengine/bevy/pull/9582#discussion_r1307517592) got me thinking, I can actually remove the bounds checks entirely. Both variants take a single `f32` value:

- `Decibels(a)` is defined for all `a`, it's just that values larger in magnitude than `800` don't really exist in `f32`, so they're all clamped to `f32::INFINITY` in amplitude. Which is defined and sound behaviour.
- `Amplitude(a)` _does_ actually exist for all `a`, if you consider the phase of the wave, not just its amplitude. A negative `a` indicates a 180° rotation of the phase, or a phase inversion. However, since nothing in the Bevy audio library works with phasing (at least that I can find), this information is largely meaningless and can just be ignored. 

So I've adjusted the access and comparison methods to reflect the true statement that the amplitude of a `Amplitude(a)` is equal to `a.abs()` for all `a`. I've also added some unit tests to document this behaviour explicitly.

hymm commented on examples/audio/audio_control.rs:
```diff
@@ -40,9 +40,9 @@ fn pause(keyboard_input: Res<Input<KeyCode>>, music_controller: Query<&AudioSink
     if let Ok(sink) = music_controller.get_single() {
         if keyboard_input.just_pressed(KeyCode::Plus) {
-            sink.set_volume(sink.volume() + 0.1);
+            sink.set_volume(VolumeLevel::Amplitude(sink.volume().amplitude() + 0.1));
```
Feels like we should implement some of the math traits for VolumeLevel and f32, so users don't have to convert types. But maybe not might get awkward since users won't always know what enum variant they're adding to.

This increased verbosity isn't great though for users wanting to do math on volume levels. Does kira have this issue?

bushrat011899 replied:
Kira doesn't implement any math operations for its [`Volume`](https://docs.rs/kira/latest/kira/enum.Volume.html) type. Personally, I think that makes sense, as the two variants are not linearly related. If this was a simple case of (for example) Feet and Metres, then the math operations would be consistent. Scaling a volume by 1.3 does completely different things depending on whether you are in decibels or amplitude, because decibels is a logarithmic scale, and amplitude is a linear one.

hymm reviewed (commented):
Do we really need 2 variants for volume? Unity, Unreal, and Godot all just use decibels. That might just be because they don't have sum types, but the current pr feels a little awkward when you want to do math on the volume.

bushrat011899 commented on crates/bevy_audio/src/volume_level.rs:
```diff
@@ -0,0 +1,196 @@
+    }
+}
+
+impl Eq for VolumeLevel {}
```
Agreed, this and `Ord` are not sensible for this type.

bushrat011899 commented:
> Do we really need 2 variants for volume? Unity, Unreal, and Godot all just use decibels. That might just be because they don't have sum types, but the current pr feels a little awkward when you want to do math on the volume.

I think using an enum is a good way to force understanding around what the volume type represents. The `Amplitude` variant is substantially more useful in engine code, since waveforms are just directly scaled by that amplitude value. Whereas, `Decibels` is more useful for user code as that unit matches real-world understanding of sound and relative "loudness".

An alternative could be to new-type `f32`, and just convert into either amplitude or decibels for all operations as required, but I don't know if that's substantially more ergonomic.

SolarLiner commented on crates/bevy_audio/src/volume_level.rs:
```diff
@@ -0,0 +1,188 @@
+    /// phase inversion. If `a` and `b` have the same magnitude (`a.abs() == b.abs()`), but opposite
+    /// sign (`a == -b`), then `Self::Amplitude(a) == Self::Amplitude(b)`, as this phase information
+    /// is currently ignored. As such, you should use positive values for this variant to avoid confusion.
+    Amplitude(f32),
```
At the risk of some bikeshedding (and probably pedantry), "amplitude" refers to the volume itself, not a specific "unit" -- `Linear` would be a better term, as it is the term that is most used for the unitless gain ratio (this applies to the method name line 26 too). I can imagine some users being confused when looking at this variant by itself and not knowing whether we are talking about the linear gain or the logarithmic one (in dB), and having to look up the enum definition to clear the confusion.

SolarLiner commented on crates/bevy_audio/src/audio.rs:
```diff
@@ -22,34 +21,12 @@ impl Default for Volume {
     /// Create a new volume level relative to the global volume.
-    pub fn new_relative(volume: f32) -> Self {
-        Self::Relative(VolumeLevel::new(volume))
+    pub fn new_relative(amplitude: f32) -> Self {
```
It's been kept as-is to not introduce breaking changes, but I do think it would be better to have `new_relative` and `new_absolute` use the `VolumeLevel`. Or we can deprecate these ones and introduce new methods (meaning 4 methods total for the cartesian product of `(relative,absolute)x(amplitude,decibel)`?
